python/across_api/base/constraints.py

```diff
@@ -0,0 +1,181 @@
+    polygon
+        Shapely Polygon object defining the SAA polygon.
+    """
+
+    polygon: Polygon
+
+    def __init__(self, polygon: list):
+        self.polygon = Polygon(polygon)
+
+    def __call__(
+        self, time: Time, ephem: EphemBase, skycoord: Optional[SkyCoord] = None
+    ) -> Union[bool, np.ndarray]:
+        """
+        Return a bool array indicating whether the spacecraft is in constraint
+        for a given ephemeris.
+
+        Arguments
+        ---------
+        time
+            The time(s) to calculate the constraint for.
+        ephem
+            The spacecraft ephemeris, must be precalculated. NOTE: The
+            ephemeris can be calculated for a longer time range than the `time`
+            argument, but it must contain the time(s) in the `time` argument.
+
+        Returns
+        -------
+            Array of booleans for every value in `time` returning True if the
+            spacecraft is in the SAA polygon at that time. If `time` is a
+            scalar then a single boolean is returned.
+        """
+
+        # Find a slice what the part of the ephemeris that we're using
+        i = getslice(time, ephem)
+
+        inconstraint = self.polygon.contains(
+            points(ephem.longitude[i], ephem.latitude[i])
+        )
+        # Return the result as True or False, or an array of True/False
+        return inconstraint[0] if time.isscalar else inconstraint
+
+
+class EarthLimbConstraint(Constraint):
+    """
+    For a given Earth limb avoidance angle, is a given coordinate inside this
+    constraint?
+
+    Parameters
+    ----------
+    min_angle
+        The minimum angle from the Earth limb that the spacecraft can point.
+
+    Methods
+    -------
+    __call__(coord, ephem, earthsize=None)
+        Checks if a given coordinate is inside the constraint.
+
+    """
+
+    min_angle: u.Quantity
+
+    def __init__(self, min_angle: u.Quantity):
+        self.min_angle = Angle(min_angle)
+
+    def __call__(
+        self,
+        time: Time,
+        ephem: EphemBase,
+        skycoord: Optional[SkyCoord] = None,
+    ) -> Union[bool, np.ndarray]:
+        """
+        Check for a given time, ephemeris and coordinate if positions given are
+        inside the Earth limb constraint. This is done by checking if the
+        separation between the Earth and the spacecraft is less than the
+        Earth's angular radius plus the minimum angle.
+
+        NOTE: Assumes a circular approximation for Earth.
+
+        Parameters
+        ----------
+        coord
+            The coordinate to check.
+        time
+            The time to check.
+        ephem
+            The ephemeris object.
+
+        Returns
+        -------
+        bool
+            `True` if the coordinate is inside the constraint, `False`
+            otherwise.
+
+        """
+        # This class requires a SkyCoord object to be passed
+        assert skycoord is not None, "SkyCoord object must be passed"
```

Comment:
If it must not be None, then why does the type signature permit it to be None?

Reply:
I'm universally defining all constraint classes through an ABC as taking skycoord as an argument. Some (e.g. SAA) can ignore it and safely have it set to `None`, others where like `EarthConstraint` require it. In the ABC you have to set it to something that will work for every class, so it's `Optional[SkyCoord]`. This assert both makes sure that you pass coordinates when it's needed, plus it lets `mypy` know that you're not passing a `None`.

Another way of handling this would just be to remove the ABC, removing the restrictions on Constraint Class structure. This would break the concept behind these classes, but not the BurstCube MVP.

Reply:
I thought that inherited methods are allowed to narrow types?

Reply:
`mypy` thinks differently.
```
base/constraints.py:135: error: Signature of "__call__" incompatible with supertype "Constraint"  [override]
base/constraints.py:135: note:      Superclass:
base/constraints.py:135: note:          def __call__(self, time: Any, ephem: EphemBase, skycoord: Any | None = ...) -> bool | ndarray[Any, Any]
base/constraints.py:135: note:      Subclass:
base/constraints.py:135: note:          def __call__(self, time: Any, ephem: EphemBase, skycoord: Any) -> bool | ndarray[Any, Any]
Found 1 error in 1 file (checked 25 source files)
```

Reply:
Anyway, [ed41cbb](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1908/commits/ed41cbb8f645194fb73d2179b73bb5a289d271e1) removes the ABC stuff. Arguments for the two constraints are now appropriate for the calls. This is good enough for the MVP we're building.